Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+      } while (anythingChanged);
+    }
+
+    void SolveAllTypeConstraints(string printableContext) {
+      PrintTypeInferenceState(printableContext);
+      PartiallySolveTypeConstraints(null);
+
+      PartiallySolveTypeConstraints(null, true);
+
+      if (TryApplyDefaultAdvice()) {
+        PartiallySolveTypeConstraints(null, true);
+      }
+
+      PrintLegend();
+      ConfirmTypeConstraints();
+      ClearState();
+    }
+
+    void ClearState() {
+      unnormalizedSubtypeConstraints.Clear();
+      equalityConstraints.Clear();
+      guardedConstraints.Clear();
+      defaultAdvice.Clear();
+      confirmations.Clear();
+      PreTypeResolver.allPreTypeProxies.Clear();
+    }
+
+    public void PrintTypeInferenceState(string/*?*/ header = null) {
+      if (!options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        return;
+      }
+      options.OutputWriter.WriteLine("*** Type inference state ***{0}", header == null ? "" : $" {header} ");
+      PrintList("Subtype constraints", unnormalizedSubtypeConstraints, stc => {
+        return $"{stc.Super} :> {stc.Sub}";
+      });
+      PrintList("Equality constraints", equalityConstraints, eqc => {
+        return $"{eqc.A} == {eqc.B}";
+      });
+#if IS_THERE_A_GOOD_WAY_TO_PRINT_GUARDED_CONSTRAINTS
+      PrintList("Guarded constraints", guardedConstraints, gc => {
+        return gc.Kind + Util.Comma("", gc.Arguments, arg => $" {arg}");
+      });
+#endif
+      PrintList("Default-type advice", defaultAdvice, advice => {
+        return $"{advice.PreType} ~-~-> {advice.WhatString}";
+      });
+#if IS_THERE_A_GOOD_WAY_TO_PRINT_CONFIRMATIONS
+      PrintList("Post-inference confirmations", confirmations, c => {
+        return $"{TokToShortLocation(c.tok)}: {c.Check} {c.PreType}: {c.ErrorMessage()}";
+      });
+#endif
+    }
+
+    void PrintLegend() {
+      PrintList("Legend", PreTypeResolver.allPreTypeProxies, pair => {
+        var s = Pad($"?{pair.Item1.UniqueId}", 4) + pair.Item1;
+        return pair.Item2 == null ? s : $"{Pad(s, 20)}  {pair.Item2}";
+      });
+    }
+
+    void PrintList<T>(string rubric, List<T> list, Func<T, string> formatter) {
+      if (!options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        return;
+      }
+      options.OutputWriter.WriteLine($"    {rubric}:");
+      foreach (var t in list) {
+        var info = $"        {formatter(t)}";
+        if (t is PreTypeStateWithErrorMessage preTypeStateWithErrorMessage) {
+          info = $"{Pad(info, 30)}  {TokToShortLocation(preTypeStateWithErrorMessage.tok)}: {preTypeStateWithErrorMessage.ErrorMessage()}";
+        }
+        options.OutputWriter.WriteLine(info);
+      }
+    }
+
+    public void DebugPrint(string format, params object[] args) {
+      if (options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        Console.WriteLine(format, args);
+      }
+    }
+
+    // ---------------------------------------- Equality constraints ----------------------------------------
```

Comment:
I find these topic headers to be a code smell. If you really need them then the code should probably in a separate class. I think you should remove them in this file.